### ISSUE TITLE
Fix or Remove Broken Local Search for MIO

### ISSUE THREAD

**Issue by schweikl** [VoglSebastian/evosuite#16]
This issue hampers our progress on #6. We must make a decision how to resolve it.
There are two choices:
 1. ~Remove Local Search for MIO because it's already broken in the `master` branch of EvoSuite.~
 2. **Attempt to fix it.** (chosen solution)

### Context
Using the version of EvoSuite from the **`master` branch (!)**, applying Local Search together with MIO fails with a `ClassCastException`:
```
java.lang.ClassCastException: org.evosuite.testcase.TestChromosome cannot be cast to org.evosuite.testsuite.TestSuiteChromosome
```

### Steps to Reproduce
1. Checkout the `master` branch of EvoSuite
2. Download the SF110 test corpus
3. Run EvoSuite 

### EvoSuite Arguments
Make sure your working directory is `47_dvd-homevideo`. Then, run EvoSuite with the following command line arguments:
 - `-class Menu`
 - `-Dalgorithm=MIO -Dstrategy=MOSuite -generateMOSuite` (to enable `MIO`)
 - `-Dlocal_search_rate=1` (to apply local search after every generation)

### Current Result
EvoSuite crashes unexpectedly with a `ClassCastException`. This happens with both JDK 1.8 and JDK 11. Full stack trace:
```
* EvoSuite 
* Going to generate test cases for class: Menu
* Starting Client-0
* Properties loaded from /home/schweikl/.../SF110-Sources/47_dvd-homevideo/evosuite-files/evosuite.properties
* Connecting to master process on port 6198
* Analyzing classpath: 
* Inheritance tree loaded from evosuite-files/inheritance.xml.gz
* Finished analyzing classpath
* Generating tests for class Menu
* Test criteria:
  - Line Coverage
  - Branch Coverage
  - Exception
  - Mutation testing (weak)
  - Method-Output Coverage
  - Top-Level Method Coverage
  - No-Exception Top-Level Method Coverage
  - Context Branch Coverage
* Total number of test goals for MIO: 423
* Using seed 1597232375087
* Starting evolution
[Progress:>                             0%] [Cov:======>                            18%]* Computation finished
[MASTER] 13:39:39.297 [logback-2] ERROR ClientNodeImpl - Error when generating tests for: Menu with seed 1597232375087. Configuration id : null
java.lang.ClassCastException: org.evosuite.testcase.TestChromosome cannot be cast to org.evosuite.testsuite.TestSuiteChromosome
	at org.evosuite.ga.metaheuristics.MIO.generateSolution(MIO.java:183) ~[classes/:na]
	at org.evosuite.strategy.MOSuiteStrategy.generateTests(MOSuiteStrategy.java:113) ~[classes/:na]
	at org.evosuite.TestSuiteGenerator.generateTests(TestSuiteGenerator.java:681) ~[classes/:na]
	at org.evosuite.TestSuiteGenerator.generateTestSuite(TestSuiteGenerator.java:245) ~[classes/:na]
	at org.evosuite.rmi.service.ClientNodeImpl.lambda$startNewSearch$0(ClientNodeImpl.java:155) ~[classes/:na]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_262]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_262]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_262]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_262]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_262]
[MASTER] 13:39:39.406 [main] ERROR SearchStatistics - No statistics has been saved because EvoSuite failed to generate any test case
[MASTER] 13:39:39.506 [main] ERROR TestGeneration - failed to write statistics data

Process finished with exit code 0
```
 
### Expected result
EvoSuite should not crash. I believe this is a bug in the program rather than a user error. To this, the type error should either be fixed or the offending lines of code should be removed altogether.

### Additional info
To reiterate, this issue is already present in the **`master`** branch of EvoSuite. The root cause for the exception is this line:
https://github.com/VoglSebastian/evosuite/blob/ae735ca5a00d3c7092d12151d9358dbe513077f3/client/src/main/java/org/evosuite/ga/metaheuristics/MIO.java#L183
Here, `this.population.get(0)` returns a `TestChromosome`, and a subsequent unchecked cast from `TestChromosome` to `TestSuiteChromosome` is attempted, which fails as both types are siblings in the inheritance hierarchy. There is no compile-time warning about this problem due to the misuse of generics to circumvent the type system.

This issue is also described in #6. In the `selfType` branch the misuse of generics was addressed. Now, line 183 (as shown above) contains a type error which is detected by the compiler (BTW, this is exactly what you want to have – type errors should be detected at compile time). But for now, line 183 is commented out in the `selfType` branch to make EvoSuite compile again. Yet, it is unclear how to address this issue. A decision has to be made whether Local Search should be fixed or removed (since it's already broken in the `master` branch).

**Comment by schweikl** [comment]
@gofraser Could you please weigh in and give us your opinion on this?

**Comment by gofraser** [comment]
I don't think we need to invest resources into fixing local search for MIO. Is the alternative you are suggesting to remove LS from MIO, or remove LS altogether?

**Comment by schweikl** [comment]
Sorry for not expressing myself more clearly. Removing LS from MIO would be sufficient. The remaining implementation of LS can stay since it doesn't cause any problems, as far as I am concerned. Alternatively, PR #18 seems like an easy fix to make LS work for MIO.

**Comment by gofraser** [comment]
I would be fine with dropping it, but since there's already #18 I suppose the problem is solved by merging that? Both options are fine for me.

**Comment by schweikl** [comment]
I'd like to give it a try whether #18 actually fixes the problem by executing EvoSuite and seeing what happens. But to do this, we must close parts of #5 first...

**Comment by schweikl** [comment]
Fixed by #18
